<a href="https://colab.research.google.com/github/WiratchawaKannika/DPDM65/blob/main/preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

In [2]:
from google.colab import drive ## import library == Google Drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
project_df = pd.read_csv('/content/drive/MyDrive/DPDM65/opendata_project.csv')
print(project_df.shape)  ##บอกขนาดของ DataFrame
project_df.head()

(23434, 45)


row_number                project_id                             name_en  \
0      7880.0  5ee08d64dda791001299b2a6  Winning Residence Sukhumvit Praksa   
1     16689.0             project_11326                    Gift Condominium   
2       290.0  62739c4089233e001362b001    Sanrak 9 Sukprayoon Nong Krathum   
3     20719.0              project_2010                     Baan Sansabai 1   
4       199.0             project_16838          Baanpruksa Ratchapruek-345   

                              name_th propertytype_id propertytype_name_en  \
0  วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1       Detached House   
1                 กิ๊ฟท์ คอนโดมิเนียม               2                Condo   
2      แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3             Townhome   
3                       บ้านแสนสบาย 1               1       Detached House   
4              บ้านพฤกษา ราชพฤกษ์-345               3             Townhome   

  propertytype_name_th  price_min     developer_id  \
0                 บ้าน  3790000.0  developer_10678   
1                คอนโด  1100000.0   developer_7319   
2             ทาวน์โฮม        0.0  developer_10510   
3                 บ้าน        0.0  developer_10510   
4             ทาวน์โฮม  1930000.0   developer_1026   

              developer_name_en  ... facility_meeting  facility_park  \
0  Winning Asset Plus Co., Ltd.  ...              NaN            NaN   
1      Gift Asset Land Co.,Ltd.  ...              NaN            1.0   
2                           NaN  ...              NaN            NaN   
3                           NaN  ...              NaN            NaN   
4  Preuksa Real Estate Co.,Ltd.  ...              NaN            1.0   

   facility_playground facility_pool facility_security         date_created  \
0                  NaN           1.0               1.0  2020-06-10 14:36:04   
1                  NaN           1.0               1.0  2017-10-30 15:38:48   
2                  NaN           NaN               NaN  2022-05-05 16:43:28   
3                  NaN           NaN               NaN  2016-09-05 15:32:15   
4                  NaN           1.0               1.0  2019-09-05 11:49:37   

           date_finish         date_updated  source  \
0                  NaN  2020-12-04 10:32:58  baania   
1  2013-01-01 07:00:00  2020-12-04 10:49:00  baania   
2                  NaN  2022-05-05 17:30:03  baania   
3  1970-01-02 07:00:00  2020-12-04 10:45:31  baania   
4  1970-01-02 07:00:00  2021-03-17 12:00:04  baania   

                                         url_project  
0  https://baania.com/th/project/วินนิ่ง-เรสซิเดน...  
1  https://baania.com/th/project/กิ๊ฟท์-คอนโดมิเน...  
2  https://baania.com/th/project/แสนรัก-9-ศุขประย...  
3  https://baania.com/th/project/บ้านแสนสบาย-1-pr...  
4  https://baania.com/th/project/บ้านพฤกษา-ราชพฤก...  

[5 rows x 45 columns]

In [4]:
unit_df = pd.read_csv('/content/drive/MyDrive/DPDM65/opendata_unittype.csv')
print(unit_df.shape)  ##บอกขนาดของ DataFrame
unit_df.head()

(42625, 27)


row_number     unittype_id    project_id             name_th  \
0       27501   unittype_9565  project_2446           2 Bedroom   
1       28777  unittype_10267  project_3543  บ้านแฝดสองชั้น Z-2   
2       28865    unittype_471   project_143    1 Bedroom Type B   
3       24988  unittype_19506  project_7766    Studio Floor 1-2   
4       38596  unittype_10290  project_3551                TH 3   

   propertytype_id propertytype_name_en propertytype_name_th  \
0                2                Condo                คอนโด   
1            20000           Twin house              บ้านแฝด   
2                2                Condo                คอนโด   
3                2                Condo                คอนโด   
4                3             Townhome             ทาวน์โฮม   

   area_total_min_wa  area_usable_min  price_min  ...  count_room_guest  \
0                NaN             37.3  2500000.0  ...               1.0   
1               40.0            150.0  2390000.0  ...               NaN   
2                NaN             28.0        NaN  ...               NaN   
3                NaN             27.0   170000.0  ...               NaN   
4                NaN            182.0  2390000.0  ...               1.0   

   count_room_kitchen  count_room_living  count_room_maid  count_room_storage  \
0                 1.0                NaN              NaN                 NaN   
1                 NaN                NaN              NaN                 NaN   
2                 NaN                NaN              NaN                 NaN   
3                 NaN                NaN              NaN                 NaN   
4                 1.0                NaN              NaN                 2.0   

   count_room_utility         date_created         date_updated  source  \
0                 NaN  2021-11-19 18:00:09  2021-11-19 17:44:45  baania   
1                 NaN  2022-01-27 14:30:04  2022-01-27 14:22:55  baania   
2                 NaN  2022-01-14 15:00:10  2022-01-14 14:40:04  baania   
3                 NaN  2020-12-04 10:40:59  2020-10-14 01:06:48  baania   
4                 NaN  2021-11-12 13:30:08  2021-02-03 12:12:54  baania   

                                        url_unittype  
0  https://baania.com/th/project/เฟ-ลิค-คอนโด-ลาด...  
1  https://baania.com/th/project/เซน-บาย-ศิวาลัย-...  
2  https://baania.com/th/project/วันพลัส-นายน์ทีน...  
3  https://baania.com/th/project/บ้านสวนชมน้ำ-776...  
4  https://baania.com/th/project/เบล็สทาวน์-ศรีนค...  

[5 rows x 27 columns]

# Data Cleaning

## Dealing with missing data

### Check missing data

In [5]:
project_df.describe()

row_number      latitude     longitude  subdistrict_id   district_id  \
count  23429.000000  23434.000000  23429.000000    23415.000000  23424.000000   
mean   11715.000000     16.399159    100.600086     4216.201751   4001.917435   
std     6763.514064    176.896370      0.947247     2476.968321   1662.485878   
min        1.000000      6.839354     97.985808     1007.000000   1006.000000   
25%     5858.000000     13.347151    100.400282     3414.000000   3411.000000   
50%    11715.000000     13.753725    100.601281     3802.000000   3800.000000   
75%    17572.000000     13.946742    100.912703     4527.000000   4000.000000   
max    23429.000000  12160.000000    105.222114    20688.000000  20003.000000   

        province_id        zipcode    count_unit  count_unittype  \
count  23425.000000   23403.000000  21547.000000    18603.000000   
mean    3942.592231   28209.221211    199.043765        2.325861   
std     1663.662562   24724.177495    320.029641        2.138378   
min        1.000000       1.000000      1.000000        1.000000   
25%     3372.000000   10400.000000     49.000000        1.000000   
50%     3781.000000   20000.000000    108.000000        2.000000   
75%     3781.000000   40000.000000    240.000000        3.000000   
max    10508.000000  390000.000000  10074.000000       96.000000   

       facility_clubhouse  facility_fitness  facility_meeting  facility_park  \
count         6820.000000       8895.000000       2890.000000   11675.000000   
mean             0.724340          0.848679          0.224913       0.894304   
std              0.446879          0.358382          0.417598       0.307461   
min              0.000000          0.000000          0.000000       0.000000   
25%              0.000000          1.000000          0.000000       1.000000   
50%              1.000000          1.000000          0.000000       1.000000   
75%              1.000000          1.000000          0.000000       1.000000   
max              1.000000          1.000000          1.000000       1.000000   

       facility_playground  facility_pool  facility_security  
count          6277.000000    9718.000000       16123.000000  
mean              0.697308       0.874048           0.959809  
std               0.459460       0.331812           0.196413  
min               0.000000       0.000000           0.000000  
25%               0.000000       1.000000           1.000000  
50%               1.000000       1.000000           1.000000  
75%               1.000000       1.000000           1.000000  
max               1.000000       1.000000           1.000000

- "count" นับเฉพาะช่องที่ไม่มี missing เท่านั้น 
> "facility_clubhouse" มีแค่ 6820 rosw จากทั้งหมด 23434 rosw

In [6]:
project_df.isnull()

row_number  project_id  name_en  name_th  propertytype_id  \
0           False       False    False    False            False   
1           False       False    False    False            False   
2           False       False    False    False            False   
3           False       False    False    False            False   
4           False       False    False    False            False   
...           ...         ...      ...      ...              ...   
23429       False       False    False    False            False   
23430       False       False    False    False            False   
23431       False       False    False    False            False   
23432       False       False    False    False            False   
23433       False       False    False    False            False   

       propertytype_name_en  propertytype_name_th  price_min  developer_id  \
0                     False                 False      False         False   
1                     False                 False      False         False   
2                     False                 False      False         False   
3                     False                 False      False         False   
4                     False                 False      False         False   
...                     ...                   ...        ...           ...   
23429                 False                 False      False         False   
23430                 False                 False      False         False   
23431                 False                 False      False         False   
23432                 False                 False      False         False   
23433                 False                 False      False         False   

       developer_name_en  ...  facility_meeting  facility_park  \
0                  False  ...              True           True   
1                  False  ...              True          False   
2                   True  ...              True           True   
3                   True  ...              True           True   
4                  False  ...              True          False   
...                  ...  ...               ...            ...   
23429              False  ...              True          False   
23430              False  ...              True           True   
23431              False  ...              True          False   
23432              False  ...              True          False   
23433               True  ...              True           True   

       facility_playground  facility_pool  facility_security  date_created  \
0                     True          False              False         False   
1                     True          False              False         False   
2                     True           True               True         False   
3                     True           True               True         False   
4                     True          False              False         False   
...                    ...            ...                ...           ...   
23429                False           True              False         False   
23430                 True           True              False         False   
23431                False           True              False         False   
23432                False          False              False         False   
23433                 True           True               True         False   

       date_finish  date_updated  source  url_project  
0             True         False   False        False  
1            False         False   False        False  
2             True         False   False        False  
3            False         False   False        False  
4            False         False   False        False  
...            ...           ...     ...          ...  
23429        False         False   False        False  
23430        False         False   False        False  
23431        False 

```
.any() 
```
เอาค่าความจริงใน Columns มา or กัน


In [7]:
project_df.isnull().any()

row_number                 True
project_id                False
name_en                   False
name_th                   False
propertytype_id           False
propertytype_name_en      False
propertytype_name_th      False
price_min                  True
developer_id              False
developer_name_en          True
developer_name_th          True
latitude                  False
longitude                  True
neighborhood_id            True
neighborhood_name_en       True
neighborhood_name_th       True
subdistrict_id             True
subdistrict_name_en        True
subdistrict_name_th        True
district_id                True
district_name_en           True
district_name_th           True
province_id                True
province_name_en           True
province_name_th           True
zipcode                    True
count_elevator             True
count_elevator_service     True
count_floor                True
count_parking              True
count_tower                True
count_un



```
.any().any()
```

บอกว่าใน DataFrame นี้มีค่า missing อยู่ 

In [8]:
project_df.isnull().any().any()

True



```
.all()
```

นำค่าความจริงใน columns มา and กัน 


In [9]:
project_df.isnull().all()

row_number                False
project_id                False
name_en                   False
name_th                   False
propertytype_id           False
propertytype_name_en      False
propertytype_name_th      False
price_min                 False
developer_id              False
developer_name_en         False
developer_name_th         False
latitude                  False
longitude                 False
neighborhood_id           False
neighborhood_name_en      False
neighborhood_name_th      False
subdistrict_id            False
subdistrict_name_en       False
subdistrict_name_th       False
district_id               False
district_name_en          False
district_name_th          False
province_id               False
province_name_en          False
province_name_th          False
zipcode                   False
count_elevator            False
count_elevator_service    False
count_floor               False
count_parking             False
count_tower               False
count_un

### delete records with missing data 

```
dropna()
```


- null
- Nan (Not a number)
- Na (Not  authorised/ Not available) == missing


In [10]:
project_df.shape

(23434, 45)

In [11]:
unit_df.shape

(42625, 27)

In [12]:
project_df.dropna()

Empty DataFrame
Columns: [row_number, project_id, name_en, name_th, propertytype_id, propertytype_name_en, propertytype_name_th, price_min, developer_id, developer_name_en, developer_name_th, latitude, longitude, neighborhood_id, neighborhood_name_en, neighborhood_name_th, subdistrict_id, subdistrict_name_en, subdistrict_name_th, district_id, district_name_en, district_name_th, province_id, province_name_en, province_name_th, zipcode, count_elevator, count_elevator_service, count_floor, count_parking, count_tower, count_unit, count_unittype, facility_clubhouse, facility_fitness, facility_meeting, facility_park, facility_playground, facility_pool, facility_security, date_created, date_finish, date_updated, source, url_project]
Index: []

[0 rows x 45 columns]

In [13]:
unit_df.dropna()

Empty DataFrame
Columns: [row_number, unittype_id, project_id, name_th, propertytype_id, propertytype_name_en, propertytype_name_th, area_total_min_wa, area_usable_min, price_min, count_airconditioner, count_elevator, count_floor_internal, count_parking, count_room_bath, count_room_bed, count_room_dinning, count_room_guest, count_room_kitchen, count_room_living, count_room_maid, count_room_storage, count_room_utility, date_created, date_updated, source, url_unittype]
Index: []

[0 rows x 27 columns]

### pandas.DataFrame.dropna

`DataFrame.dropna(axis=0, how='any', thresh=None, subset=None, inplace=False)`

- axis (default = 0) ต้องการสนใจในแนว rows หรือ columns
- thresh แถวนั้นจะถูก drop ถ้ามีจำนวน Na น้อยกว่าจำนวนที่เรา set ไว้ 
- subset คือ ระบุ columns ที่ต้องการ dropna() เป็น list ของ Colums เช่น   ['colums1', 'columns2'] 
- inplace คือ function ที่ keep ผลของการ dropna() ไว้ในตารางนั้นเลย โดยไม่ต้องสร้างตัวแปรขึ้นมารับค่าไว้ใหม่ คือ การ update ตัวแปรเดิม หรือ ใช้เมื่อจะไม่นำตารางเดิมไปใช้ต่อ 



```
# 1. axis
```



In [14]:
project_df.dropna(axis=1)  ## Drop ทุก Columns ที่มี midssing อยู่  ## ''' เหลือแค่ 8 columns 

project_id                                       name_en  \
0      5ee08d64dda791001299b2a6            Winning Residence Sukhumvit Praksa   
1                 project_11326                              Gift Condominium   
2      62739c4089233e001362b001              Sanrak 9 Sukprayoon Nong Krathum   
3                  project_2010                               Baan Sansabai 1   
4                 project_16838                    Baanpruksa Ratchapruek-345   
...                         ...                                           ...   
23429             project_12317  Baan Klangmeung Urbanion Ladphrao-Chokchai 4   
23430             project_12057                           @Home Kaset-Nawamin   
23431              project_4461                    I-Design Wongwaen-Ramintra   
23432               project_144                 Koolpunt Ville 15 Park Avenue   
23433             project_16300                          Thawisuk Garden Home   

                                             name_th propertytype_id  \
0                 วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1   
1                                กิ๊ฟท์ คอนโดมิเนียม               2   
2                     แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3   
3                                      บ้านแสนสบาย 1               1   
4                             บ้านพฤกษา ราชพฤกษ์-345               3   
...                                              ...             ...   
23429  บ้านกลางเมือง เออร์บาเนี่ยน ลาดพร้าว-โชคชัย 4               3   
23430                          แอทโฮม เกษตร-นวมินทร์               3   
23431                     ไอ-ดีไซน์ วงแหวน-รามอินทรา               4   
23432                  กุลพันธ์วิลล์ 15 พาร์ค อเวนิว               1   
23433                             ทวีสุข การ์เด้นโฮม               1   

      propertytype_name_en propertytype_name_th     developer_id   latitude  
0           Detached House                 บ้าน  developer_10678  13.570530  
1                    Condo                คอนโด   developer_7319  13.033814  
2                 Townhome             ทาวน์โฮม  developer_10510  13.417402  
3           Detached House                 บ้าน  developer_10510  15.038042  
4                 Townhome             ทาวน์โฮม   developer_1026  13.965246  
...                    ...                  ...              ...        ...  
23429             Townhome             ทาวน์โฮม   developer_1760  13.814468  
23430             Townhome             ทาวน์โฮม   developer_1793  13.834361  
23431           Commercial         อาคารพาณิชย์   developer_2217  13.859872  
23432       Detached House                 บ้าน   developer_1004  18.808447  
23433       Detached House                 บ้าน  developer_10510  12.684945  

[23434 rows x 8 columns]

In [19]:
project_df.dropna(axis=1).isnull().any().any()

False

In [17]:
print(project_df.shape)
project_df.isnull().any().any()

(23434, 45)


True

In [21]:
project_df2 = project_df.dropna(axis=1) ## project_df2 สร้างตัวแปร project_df2 มารับค่าไว้เพื่อนำไปประมวลผลต่อได้
print(project_df2.shape)
project_df2.head()

(23434, 8)


project_id                             name_en  \
0  5ee08d64dda791001299b2a6  Winning Residence Sukhumvit Praksa   
1             project_11326                    Gift Condominium   
2  62739c4089233e001362b001    Sanrak 9 Sukprayoon Nong Krathum   
3              project_2010                     Baan Sansabai 1   
4             project_16838          Baanpruksa Ratchapruek-345   

                              name_th propertytype_id propertytype_name_en  \
0  วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1       Detached House   
1                 กิ๊ฟท์ คอนโดมิเนียม               2                Condo   
2      แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3             Townhome   
3                       บ้านแสนสบาย 1               1       Detached House   
4              บ้านพฤกษา ราชพฤกษ์-345               3             Townhome   

  propertytype_name_th     developer_id   latitude  
0                 บ้าน  developer_10678  13.570530  
1                คอนโด   developer_7319  13.033814  
2             ทาวน์โฮม  developer_10510  13.417402  
3                 บ้าน  developer_10510  15.038042  
4             ทาวน์โฮม   developer_1026  13.965246



```
# 2. thresh
```



In [22]:
project_df.iloc[:5, -7:] ## เอาแถวที่ 0-4, และ 7 Columns สุดท้าย  

facility_pool  facility_security         date_created          date_finish  \
0            1.0                1.0  2020-06-10 14:36:04                  NaN   
1            1.0                1.0  2017-10-30 15:38:48  2013-01-01 07:00:00   
2            NaN                NaN  2022-05-05 16:43:28                  NaN   
3            NaN                NaN  2016-09-05 15:32:15  1970-01-02 07:00:00   
4            1.0                1.0  2019-09-05 11:49:37  1970-01-02 07:00:00   

          date_updated  source  \
0  2020-12-04 10:32:58  baania   
1  2020-12-04 10:49:00  baania   
2  2022-05-05 17:30:03  baania   
3  2020-12-04 10:45:31  baania   
4  2021-03-17 12:00:04  baania   

                                         url_project  
0  https://baania.com/th/project/วินนิ่ง-เรสซิเดน...  
1  https://baania.com/th/project/กิ๊ฟท์-คอนโดมิเน...  
2  https://baania.com/th/project/แสนรัก-9-ศุขประย...  
3  https://baania.com/th/project/บ้านแสนสบาย-1-pr...  
4  https://baania.com/th/project/บ้านพฤกษา-ราชพฤก...

In [23]:
project_df.iloc[:5, -7:].dropna(thresh=5) ##ถ้า missing ใน rows นั้นมีมากวว่า 5 ค่า  

facility_pool  facility_security         date_created          date_finish  \
0            1.0                1.0  2020-06-10 14:36:04                  NaN   
1            1.0                1.0  2017-10-30 15:38:48  2013-01-01 07:00:00   
3            NaN                NaN  2016-09-05 15:32:15  1970-01-02 07:00:00   
4            1.0                1.0  2019-09-05 11:49:37  1970-01-02 07:00:00   

          date_updated  source  \
0  2020-12-04 10:32:58  baania   
1  2020-12-04 10:49:00  baania   
3  2020-12-04 10:45:31  baania   
4  2021-03-17 12:00:04  baania   

                                         url_project  
0  https://baania.com/th/project/วินนิ่ง-เรสซิเดน...  
1  https://baania.com/th/project/กิ๊ฟท์-คอนโดมิเน...  
3  https://baania.com/th/project/บ้านแสนสบาย-1-pr...  
4  https://baania.com/th/project/บ้านพฤกษา-ราชพฤก...



```
# 3. subset 
```



In [25]:
project_df.head()

row_number                project_id                             name_en  \
0      7880.0  5ee08d64dda791001299b2a6  Winning Residence Sukhumvit Praksa   
1     16689.0             project_11326                    Gift Condominium   
2       290.0  62739c4089233e001362b001    Sanrak 9 Sukprayoon Nong Krathum   
3     20719.0              project_2010                     Baan Sansabai 1   
4       199.0             project_16838          Baanpruksa Ratchapruek-345   

                              name_th propertytype_id propertytype_name_en  \
0  วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1       Detached House   
1                 กิ๊ฟท์ คอนโดมิเนียม               2                Condo   
2      แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3             Townhome   
3                       บ้านแสนสบาย 1               1       Detached House   
4              บ้านพฤกษา ราชพฤกษ์-345               3             Townhome   

  propertytype_name_th  price_min     developer_id  \
0                 บ้าน  3790000.0  developer_10678   
1                คอนโด  1100000.0   developer_7319   
2             ทาวน์โฮม        0.0  developer_10510   
3                 บ้าน        0.0  developer_10510   
4             ทาวน์โฮม  1930000.0   developer_1026   

              developer_name_en  ... facility_meeting  facility_park  \
0  Winning Asset Plus Co., Ltd.  ...              NaN            NaN   
1      Gift Asset Land Co.,Ltd.  ...              NaN            1.0   
2                           NaN  ...              NaN            NaN   
3                           NaN  ...              NaN            NaN   
4  Preuksa Real Estate Co.,Ltd.  ...              NaN            1.0   

   facility_playground facility_pool facility_security         date_created  \
0                  NaN           1.0               1.0  2020-06-10 14:36:04   
1                  NaN           1.0               1.0  2017-10-30 15:38:48   
2                  NaN           NaN               NaN  2022-05-05 16:43:28   
3                  NaN           NaN               NaN  2016-09-05 15:32:15   
4                  NaN           1.0               1.0  2019-09-05 11:49:37   

           date_finish         date_updated  source  \
0                  NaN  2020-12-04 10:32:58  baania   
1  2013-01-01 07:00:00  2020-12-04 10:49:00  baania   
2                  NaN  2022-05-05 17:30:03  baania   
3  1970-01-02 07:00:00  2020-12-04 10:45:31  baania   
4  1970-01-02 07:00:00  2021-03-17 12:00:04  baania   

                                         url_project  
0  https://baania.com/th/project/วินนิ่ง-เรสซิเดน...  
1  https://baania.com/th/project/กิ๊ฟท์-คอนโดมิเน...  
2  https://baania.com/th/project/แสนรัก-9-ศุขประย...  
3  https://baania.com/th/project/บ้านแสนสบาย-1-pr...  
4  https://baania.com/th/project/บ้านพฤกษา-ราชพฤก...  

[5 rows x 45 columns]

In [28]:
#project_df.columns

- เช่น สนใจแค่ราคา (price_min)  และชนิดของของโครงการ (propertytype_name_th)

In [30]:
project_df.dropna(subset=['propertytype_name_en', 'price_min'])

row_number                project_id  \
0          7880.0  5ee08d64dda791001299b2a6   
1         16689.0             project_11326   
2           290.0  62739c4089233e001362b001   
3         20719.0              project_2010   
4           199.0             project_16838   
...           ...                       ...   
23429     12883.0             project_12317   
23430     21181.0             project_12057   
23431     13993.0              project_4461   
23432      5889.0               project_144   
23433      5529.0             project_16300   

                                            name_en  \
0                Winning Residence Sukhumvit Praksa   
1                                  Gift Condominium   
2                  Sanrak 9 Sukprayoon Nong Krathum   
3                                   Baan Sansabai 1   
4                        Baanpruksa Ratchapruek-345   
...                                             ...   
23429  Baan Klangmeung Urbanion Ladphrao-Chokchai 4   
23430                           @Home Kaset-Nawamin   
23431                    I-Design Wongwaen-Ramintra   
23432                 Koolpunt Ville 15 Park Avenue   
23433                          Thawisuk Garden Home   

                                             name_th propertytype_id  \
0                 วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1   
1                                กิ๊ฟท์ คอนโดมิเนียม               2   
2                     แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3   
3                                      บ้านแสนสบาย 1               1   
4                             บ้านพฤกษา ราชพฤกษ์-345               3   
...                                              ...             ...   
23429  บ้านกลางเมือง เออร์บาเนี่ยน ลาดพร้าว-โชคชัย 4               3   
23430                          แอทโฮม เกษตร-นวมินทร์               3   
23431                     ไอ-ดีไซน์ วงแหวน-รามอินทรา               4   
23432                  กุลพันธ์วิลล์ 15 พาร์ค อเวนิว               1   
23433                             ทวีสุข การ์เด้นโฮม               1   

      propertytype_name_en propertytype_name_th  price_min     developer_id  \
0           Detached House                 บ้าน  3790000.0  developer_10678   
1                    Condo                คอนโด  1100000.0   developer_7319   
2                 Townhome             ทาวน์โฮม        0.0  developer_10510   
3           Detached House                 บ้าน        0.0  developer_10510   
4                 Townhome             ทาวน์โฮม  1930000.0   developer_1026   
...                    ...                  ...        ...              ...   
23429             Townhome             ทาวน์โฮม  5700000.0   developer_1760   
23430             Townhome             ทาวน์โฮม  3590000.0   developer_1793   
23431           Commercial         อาคารพาณิชย์  1870000.0   developer_2217   
23432       Detached House                 บ้าน  4000000.0   developer_1004   
23433       Detached House                 บ้าน  2800000.0  developer_10510   

                   developer_name_en  ... facility_meeting  facility_park  \
0       Winning Asset Plus Co., Ltd.  ...              NaN            NaN   
1           Gift Asset Land Co.,Ltd.  ...              NaN            1.0   
2                                NaN  ...              NaN            NaN   
3                                NaN  ...              NaN            NaN   
4       Preuksa Real Estate Co.,Ltd.  ...              NaN            1.0   
...                              ...  ...              ...            ...   
23429         AP (Thailand) Co,. Ltd  ...              NaN            1.0   
23430       Areeya Property Co.,Ltd.  ...              NaN            NaN   
23431  Wangthong Group Pub Co., Ltd.  ...              NaN            1.0   
23432     Homeland Service Co., Ltd.  ...              NaN            1.0   
23433                            NaN  ...              NaN            NaN   

       facility_playground facility_pool f

In [29]:
project_df.dropna(subset=['propertytype_name_en', 'price_min']).isnull().any()

row_number                 True
project_id                False
name_en                   False
name_th                   False
propertytype_id           False
propertytype_name_en      False
propertytype_name_th      False
price_min                 False
developer_id              False
developer_name_en          True
developer_name_th          True
latitude                  False
longitude                  True
neighborhood_id            True
neighborhood_name_en       True
neighborhood_name_th       True
subdistrict_id             True
subdistrict_name_en        True
subdistrict_name_th        True
district_id                True
district_name_en           True
district_name_th           True
province_id                True
province_name_en           True
province_name_th           True
zipcode                    True
count_elevator             True
count_elevator_service     True
count_floor                True
count_parking              True
count_tower                True
count_un



```
# 4. inplace 
```



In [31]:
project_df.dropna(subset=['propertytype_name_en', 'price_min'], inplace = True) ## ไม่เอาตัวแปรมารับ คือ update ตัวแปรเดิม  ##ใช้เมื่อจะไม่นำตารางเดิมไปใช้ต่อ 

# Quiz 3

1. กำจัดค่า missing โดยการตัด column ให้น้อยที่สุด โดยจำนวน records ที่ไม่มี missing เลย ยังเหลือมากกว่า 80%
2. สร้าง data ที่ค่าใน columns ['price_min', 'neighborhood_name_en', 'developer_name_en']   



In [37]:
project_df = pd.read_csv('/content/drive/MyDrive/DPDM65/opendata_project.csv')
print(project_df.shape)  ##บอกขนาดของ DataFrame
project_df.head()

(23434, 45)


row_number                project_id                             name_en  \
0      7880.0  5ee08d64dda791001299b2a6  Winning Residence Sukhumvit Praksa   
1     16689.0             project_11326                    Gift Condominium   
2       290.0  62739c4089233e001362b001    Sanrak 9 Sukprayoon Nong Krathum   
3     20719.0              project_2010                     Baan Sansabai 1   
4       199.0             project_16838          Baanpruksa Ratchapruek-345   

                              name_th propertytype_id propertytype_name_en  \
0  วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1       Detached House   
1                 กิ๊ฟท์ คอนโดมิเนียม               2                Condo   
2      แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3             Townhome   
3                       บ้านแสนสบาย 1               1       Detached House   
4              บ้านพฤกษา ราชพฤกษ์-345               3             Townhome   

  propertytype_name_th  price_min     developer_id  \
0                 บ้าน  3790000.0  developer_10678   
1                คอนโด  1100000.0   developer_7319   
2             ทาวน์โฮม        0.0  developer_10510   
3                 บ้าน        0.0  developer_10510   
4             ทาวน์โฮม  1930000.0   developer_1026   

              developer_name_en  ... facility_meeting  facility_park  \
0  Winning Asset Plus Co., Ltd.  ...              NaN            NaN   
1      Gift Asset Land Co.,Ltd.  ...              NaN            1.0   
2                           NaN  ...              NaN            NaN   
3                           NaN  ...              NaN            NaN   
4  Preuksa Real Estate Co.,Ltd.  ...              NaN            1.0   

   facility_playground facility_pool facility_security         date_created  \
0                  NaN           1.0               1.0  2020-06-10 14:36:04   
1                  NaN           1.0               1.0  2017-10-30 15:38:48   
2                  NaN           NaN               NaN  2022-05-05 16:43:28   
3                  NaN           NaN               NaN  2016-09-05 15:32:15   
4                  NaN           1.0               1.0  2019-09-05 11:49:37   

           date_finish         date_updated  source  \
0                  NaN  2020-12-04 10:32:58  baania   
1  2013-01-01 07:00:00  2020-12-04 10:49:00  baania   
2                  NaN  2022-05-05 17:30:03  baania   
3  1970-01-02 07:00:00  2020-12-04 10:45:31  baania   
4  1970-01-02 07:00:00  2021-03-17 12:00:04  baania   

                                         url_project  
0  https://baania.com/th/project/วินนิ่ง-เรสซิเดน...  
1  https://baania.com/th/project/กิ๊ฟท์-คอนโดมิเน...  
2  https://baania.com/th/project/แสนรัก-9-ศุขประย...  
3  https://baania.com/th/project/บ้านแสนสบาย-1-pr...  
4  https://baania.com/th/project/บ้านพฤกษา-ราชพฤก...  

[5 rows x 45 columns]

In [38]:
project_df.isnull().any()

row_number                 True
project_id                False
name_en                   False
name_th                   False
propertytype_id           False
propertytype_name_en      False
propertytype_name_th      False
price_min                  True
developer_id              False
developer_name_en          True
developer_name_th          True
latitude                  False
longitude                  True
neighborhood_id            True
neighborhood_name_en       True
neighborhood_name_th       True
subdistrict_id             True
subdistrict_name_en        True
subdistrict_name_th        True
district_id                True
district_name_en           True
district_name_th           True
province_id                True
province_name_en           True
province_name_th           True
zipcode                    True
count_elevator             True
count_elevator_service     True
count_floor                True
count_parking              True
count_tower                True
count_un

In [39]:
project_df.isnull().sum() ## Columns ไหนมี missing มากที่สุด  == 

row_number                    5
project_id                    0
name_en                       0
name_th                       0
propertytype_id               0
propertytype_name_en          0
propertytype_name_th          0
price_min                   117
developer_id                  0
developer_name_en         12654
developer_name_th          8745
latitude                      0
longitude                     5
neighborhood_id            4215
neighborhood_name_en       4214
neighborhood_name_th       4220
subdistrict_id               19
subdistrict_name_en          14
subdistrict_name_th          16
district_id                  10
district_name_en             10
district_name_th             10
province_id                   9
province_name_en             10
province_name_th             10
zipcode                      31
count_elevator            21543
count_elevator_service    22820
count_floor               18718
count_parking             21430
count_tower               23429
count_un

In [43]:
project_df2 = project_df.drop('count_tower,', 'count_elevator_service', axis=1)
print(project_df2.shape)
project_df2.head()

(23434, 44)


row_number                project_id                             name_en  \
0      7880.0  5ee08d64dda791001299b2a6  Winning Residence Sukhumvit Praksa   
1     16689.0             project_11326                    Gift Condominium   
2       290.0  62739c4089233e001362b001    Sanrak 9 Sukprayoon Nong Krathum   
3     20719.0              project_2010                     Baan Sansabai 1   
4       199.0             project_16838          Baanpruksa Ratchapruek-345   

                              name_th propertytype_id propertytype_name_en  \
0  วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1       Detached House   
1                 กิ๊ฟท์ คอนโดมิเนียม               2                Condo   
2      แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3             Townhome   
3                       บ้านแสนสบาย 1               1       Detached House   
4              บ้านพฤกษา ราชพฤกษ์-345               3             Townhome   

  propertytype_name_th  price_min     developer_id  \
0                 บ้าน  3790000.0  developer_10678   
1                คอนโด  1100000.0   developer_7319   
2             ทาวน์โฮม        0.0  developer_10510   
3                 บ้าน        0.0  developer_10510   
4             ทาวน์โฮม  1930000.0   developer_1026   

              developer_name_en  ... facility_meeting  facility_park  \
0  Winning Asset Plus Co., Ltd.  ...              NaN            NaN   
1      Gift Asset Land Co.,Ltd.  ...              NaN            1.0   
2                           NaN  ...              NaN            NaN   
3                           NaN  ...              NaN            NaN   
4  Preuksa Real Estate Co.,Ltd.  ...              NaN            1.0   

   facility_playground facility_pool facility_security         date_created  \
0                  NaN           1.0               1.0  2020-06-10 14:36:04   
1                  NaN           1.0               1.0  2017-10-30 15:38:48   
2                  NaN           NaN               NaN  2022-05-05 16:43:28   
3                  NaN           NaN               NaN  2016-09-05 15:32:15   
4                  NaN           1.0               1.0  2019-09-05 11:49:37   

           date_finish         date_updated  source  \
0                  NaN  2020-12-04 10:32:58  baania   
1  2013-01-01 07:00:00  2020-12-04 10:49:00  baania   
2                  NaN  2022-05-05 17:30:03  baania   
3  1970-01-02 07:00:00  2020-12-04 10:45:31  baania   
4  1970-01-02 07:00:00  2021-03-17 12:00:04  baania   

                                         url_project  
0  https://baania.com/th/project/วินนิ่ง-เรสซิเดน...  
1  https://baania.com/th/project/กิ๊ฟท์-คอนโดมิเน...  
2  https://baania.com/th/project/แสนรัก-9-ศุขประย...  
3  https://baania.com/th/project/บ้านแสนสบาย-1-pr...  
4  https://baania.com/th/project/บ้านพฤกษา-ราชพฤก...  

[5 rows x 44 columns]

In [44]:
project_df2.isnull().sum()

row_number                    5
project_id                    0
name_en                       0
name_th                       0
propertytype_id               0
propertytype_name_en          0
propertytype_name_th          0
price_min                   117
developer_id                  0
developer_name_en         12654
developer_name_th          8745
latitude                      0
longitude                     5
neighborhood_id            4215
neighborhood_name_en       4214
neighborhood_name_th       4220
subdistrict_id               19
subdistrict_name_en          14
subdistrict_name_th          16
district_id                  10
district_name_en             10
district_name_th             10
province_id                   9
province_name_en             10
province_name_th             10
zipcode                      31
count_elevator            21543
count_elevator_service    22820
count_floor               18718
count_parking             21430
count_unit                 1887
count_un

In [48]:
project_df2 = project_df.dropna(axis=1, how='any')
project_df2 

project_id                                       name_en  \
0      5ee08d64dda791001299b2a6            Winning Residence Sukhumvit Praksa   
1                 project_11326                              Gift Condominium   
2      62739c4089233e001362b001              Sanrak 9 Sukprayoon Nong Krathum   
3                  project_2010                               Baan Sansabai 1   
4                 project_16838                    Baanpruksa Ratchapruek-345   
...                         ...                                           ...   
23429             project_12317  Baan Klangmeung Urbanion Ladphrao-Chokchai 4   
23430             project_12057                           @Home Kaset-Nawamin   
23431              project_4461                    I-Design Wongwaen-Ramintra   
23432               project_144                 Koolpunt Ville 15 Park Avenue   
23433             project_16300                          Thawisuk Garden Home   

                                             name_th propertytype_id  \
0                 วินนิ่ง เรสซิเดนท์ สุขุมวิท แพรกษา               1   
1                                กิ๊ฟท์ คอนโดมิเนียม               2   
2                     แสนรัก 9 ศุขประยูร หนองกระทุ่ม               3   
3                                      บ้านแสนสบาย 1               1   
4                             บ้านพฤกษา ราชพฤกษ์-345               3   
...                                              ...             ...   
23429  บ้านกลางเมือง เออร์บาเนี่ยน ลาดพร้าว-โชคชัย 4               3   
23430                          แอทโฮม เกษตร-นวมินทร์               3   
23431                     ไอ-ดีไซน์ วงแหวน-รามอินทรา               4   
23432                  กุลพันธ์วิลล์ 15 พาร์ค อเวนิว               1   
23433                             ทวีสุข การ์เด้นโฮม               1   

      propertytype_name_en propertytype_name_th     developer_id   latitude  
0           Detached House                 บ้าน  developer_10678  13.570530  
1                    Condo                คอนโด   developer_7319  13.033814  
2                 Townhome             ทาวน์โฮม  developer_10510  13.417402  
3           Detached House                 บ้าน  developer_10510  15.038042  
4                 Townhome             ทาวน์โฮม   developer_1026  13.965246  
...                    ...                  ...              ...        ...  
23429             Townhome             ทาวน์โฮม   developer_1760  13.814468  
23430             Townhome             ทาวน์โฮม   developer_1793  13.834361  
23431           Commercial         อาคารพาณิชย์   developer_2217  13.859872  
23432       Detached House                 บ้าน   developer_1004  18.808447  
23433       Detached House                 บ้าน  developer_10510  12.684945  

[23434 rows x 8 columns]

In [40]:
project_df2 = project_df.dropna(how='all', axis=1)
print(project_df2.shape)

(23434, 45)


In [41]:
project_df22 = project_df.dropna(how='any', axis=1)
print(project_df22.shape)

(23434, 8)


- 2. สร้าง data ที่ค่าใน columns ['price_min', 'neighborhood_name_en', 'developer_name_en']   

In [49]:
project_df3  = project_df[['price_min', 'neighborhood_name_en', 'developer_name_en']]
project_df3

price_min neighborhood_name_en              developer_name_en
0      3790000.0       MANGKON NAKDEE   Winning Asset Plus Co., Ltd.
1      1100000.0               BO WIN       Gift Asset Land Co.,Ltd.
2            0.0        NONG TAMLUENG                            NaN
3            0.0     BUENG THAP CHANG                            NaN
4      1930000.0            BANG DUEA   Preuksa Real Estate Co.,Ltd.
...          ...                  ...                            ...
23429  5700000.0            LAT PHRAO         AP (Thailand) Co,. Ltd
23430  3590000.0       KASET NAWA MIN       Areeya Property Co.,Ltd.
23431  1870000.0            KHANNAYAO  Wangthong Group Pub Co., Ltd.
23432  4000000.0     SAN PU LOEI NUEA     Homeland Service Co., Ltd.
23433  2800000.0             HAT PHLA                            NaN

[23434 rows x 3 columns]

In [52]:
project_df3.isnull().sum()

price_min                 117
neighborhood_name_en     4214
developer_name_en       12654
dtype: int64

In [57]:
project_df3_ = project_df3.dropna(subset=['developer_name_en'])
print(project_df3_.shape)
project_df3_.head()

(10780, 3)


price_min neighborhood_name_en                developer_name_en
0  3790000.0       MANGKON NAKDEE     Winning Asset Plus Co., Ltd.
1  1100000.0               BO WIN         Gift Asset Land Co.,Ltd.
4  1930000.0            BANG DUEA     Preuksa Real Estate Co.,Ltd.
5  1480000.0          RAM IN THRA         Asset Wise Pub Co., Ltd.
6   990000.0                  NaN  Sedsiri Property 2499 Co., Ltd.

In [55]:
project_df3_.isnull().sum()

price_min                 22
neighborhood_name_en    1568
developer_name_en          0
dtype: int64